Document model V0 + UI/Engine Syncing with Redux + Project structuring

tonyketcham opened the pull request:
- Adds Redux as the central state manager
- Seeds the design of the document model
  - initial design in docs/audio-graph.md
    - https://whimsical.com/project-HNvThW62PrN9Gp33UfAuRf@2Ux7TurymNHcs4gAfi4M
- Adds TS and Rust CI linting jobs

tonyketcham committed:
init redux

tonyketcham committed:
first audio engine design iteration

tonyketcham committed:
Update and rename audioEngine.md to audio-graph.md

nbennett320 commented on package.json:
```diff
@@ -43,13 +44,15 @@
     "jest-environment-jsdom": "27.5.1",
     "jest-resolve": "27.4.2",
     "jest-watch-typeahead": "1.0.0",
+    "nanoid": "^3.3.2",
```
what's nanoid/where is it used and what for?

tonyketcham replied:
a better uuid with similar chance of collision (incredibly low) - we can use it in Redux reducers as the ID field when creating new entities

tonyketcham commented on docs/audio-graph.md:
```diff
@@ -0,0 +1,23 @@
+
+### Project
+
+The highest level in the model. This includes filename, metadata, and a main output channel (ch0).
```
This could also include tempo but we need to think about how we want to allow tempo to be modulated & how that should fit in to other forms of automation within a `DeviceTrack`

Commit:
connecting backend state to frontend with redux

- implementing redux slices for backend functions
- creating basic input components
- start topbar component
- move functionality and state out of App.tsx
- make eslint less annoying
- move a lot of deps to devDependencies in package.json
- temporarily removing nanoid from dependencies (at least until needed)
- cleaning up App.tsx a lot
- implementing runtime formatting functions
- add util module

 Changes committed:
	modified:   .eslintrc.cjs
	modified:   backend/src/main.rs
	new file:   backend/src/util/format.rs
	new file:   backend/src/util/mod.rs
	modified:   package.json
	modified:   src/App.tsx
	new file:   src/components/common/Button.tsx
	new file:   src/components/common/Input.tsx
	new file:   src/components/common/base/ButtonBase.tsx
	new file:   src/components/common/base/InputBase.tsx
	new file:   src/components/common/base/styles.module.scss
	new file:   src/components/common/styles.module.scss
	new file:   src/components/topbar/PlayPauseButton.tsx
	new file:   src/components/topbar/TempoInput.tsx
	new file:   src/components/topbar/Topbar.tsx
	new file:   src/components/topbar/TopbarBase.tsx
	new file:   src/components/topbar/TopbarTimer.tsx
	new file:   src/hooks/redux.ts
	modified:   src/index.tsx
	deleted:    src/reportWebVitals.ts
	new file:   src/state/slices/playPausePlaylistSlice.ts
	new file:   src/state/slices/playlistMetronomeEnabledSlice.ts
	new file:   src/state/slices/playlistRuntimeSlice.ts
	new file:   src/state/slices/playlistTempoSlice.ts
	modified:   src/state/store.ts
	modified:   tsconfig.json

tonyketcham commented on backend/src/util/format.rs:
```diff
@@ -0,0 +1,18 @@
+use chrono;
+
+pub fn format_playlist_runtime(start_ms: i64) -> String {
+  let now_ts = chrono::offset::Local::now().timestamp();
+  let dur = now_ts - start_ms;
+  
+  let format = || -> String {
+    let m = (dur as f32 / 60_000.).floor() as i64;
+    let s = ((dur - (m * 60_000)) as f32 / 1_000.).floor() as i64;
+    let ms = dur - (m * 60_000) - (s * 1_000);
+    let mbufc: &str = if m < 10 { "0" } else { "" };
+    let sbufc: &str = if s < 10 { "0" } else { "" };
+    let msbufc: &str = if ms < 10 { "0" } else { "" };
+    format!("{sbufc}{m}:{mbufc}{s}:{msbufc}{ms}")
+  };
+
+  format()
+}
```
in the future we might want to do this kind of UI-specific unit formatting in TypeScript and keep all the data sent from the engine in `ms`

tonyketcham replied:
should be fine for now since there's no interactivity with the playtime as it's purely a computed property based on where the playhead is

tonyketcham commented on src/state/slices/playPausePlaylistSlice.ts:
```diff
@@ -0,0 +1,35 @@
+export const { setPlaying } = playPausePlaylistSlice.actions
+
+export const togglePlay = () => (dispatch: Dispatch) => {
+  invoke('toggle_playlist', {}).then(() => {
+    invoke('get_playlist_playing', {}).then((data) => {
+      dispatch(setPlaying(data as boolean))
+    })
+  })
```
can we change these to async/await?

nbennett320 replied:
yeah, fire idea, i for some reason i thought async was slower but turns out its actually easier to implement optimizations with async/await... done

tonyketcham commented on src/state/slices/playlistMetronomeEnabledSlice.ts:
```diff
@@ -0,0 +1,36 @@
+export const { setEnabled } = playlistMetronomeEnabled.actions
+
+export const toggleMetronome = () => (dispatch: Dispatch) => {
+  invoke('toggle_metronome_enabled', {}).then(() => {
+    invoke('get_metronome_enabled', {}).then((data) => {
+      dispatch(setEnabled(data as boolean))
+    })
+  })
```
same here as well as in the other slices invoking Tauri-wrapped methods

nbennett320 replied:
aye aye

tonyketcham commented on src/state/store.ts:
```diff
@@ -0,0 +1,20 @@
+
+export const store = configureStore({
+  reducer: {
+    playlistPlaying: playPausePlaylistReducer,
+    playlistTempo: playlistTempoReducer,
+    playlistMetronomeEnabled: playlistMetronomeEnabledReducer,
+    playlistRuntime: playlistRuntimeReducer
```
These all seem like the same category of state - can we merge these into one compositional slice? This level of granularity will become hard to maintain at scale

nbennett320 replied:
Yea that’s a fire idea, wasn’t sure how to structure it

nbennett320 replied:
this was actually the best idea ever... just got it done

tonyketcham commented on src/components/topbar/TopbarTimer.tsx:
```diff
@@ -0,0 +1,39 @@
+import { 
+  selectPlaylistPlaying,
+  PlayPausePlaylistState
+} from '../../state/slices/playPausePlaylistSlice'
```
we could set the `baseUrl = "src"` in `tsconfig` to avoid having to do any `../../` action 😉

nbennett320 replied:
does this need any vite configuration..? i think with webpack it might have been needed (?)

tonyketcham commented on src/components/common/base/ButtonBase.tsx:
```diff
@@ -0,0 +1,18 @@
+import React from 'react'
+import styles from './styles.module.scss'
+
+export interface Props<T> extends React.HTMLProps<T> {
+  type: 'button' | 'reset' | 'submit',
+}
+
+const ButtonBase = (props: Props<HTMLButtonElement>) => (
+  <button
+    {...props}
+    className={`input ${styles.ButtonBase}`}
+    type={props?.type ?? 'button'}
+  >
+    {props.children} 
+  </button>
+)
+
+export default ButtonBase
```
nice call on these base components

nbennett320 replied:
ty :D

Commit:
committing pnpm-lock / fix build pipeline

Commit:
consolidating playlist slices to one file, clean up, move then catch functions to async

Commit:
release build optimizations in cargo.toml

Commit:
add script to link assets to target directory in package.json

Commit:
init resources page

Commit:
adding resources

nbennett320 commented:
can we merge 💀

tonyketcham committed:
Add linter jobs to CI

tonyketcham committed:
add rustfmt

tonyketcham committed:
update cargo meta

tonyketcham committed:
run formatter

tonyketcham committed:
set working dir for clippy job

tonyketcham commented:
Still trying to get Clippy workin in CI... might need to bounce it out to a separate PR. in the meantime run:
`cargo clippy && cargo check`

tonyketcham commented:
> can we merge 💀

yeah give it one last rinse but I think we've beaten this thing to death and can just follow up w/ more PRs to correct Types, etc.

Commit:
changing fmt width to 80

Commit:
update lockfile
